### PR TITLE
 Disallow ferries to swim on lava 

### LINKED ISSUE #4173: Ferrys trying to reach their waterway they want to serve can get stuck on lava-water-boundaries.
### Describe the bug
Ferrys trying to reach their waterway they want to serve can get stuck on lava-water-boundaries:  
Waterways can be build along "spikes" of land reaching into water (i.e. where the waterway touches, but does not pass along, land). Touching lava is not possible.  
However, when a waterway is build elsewhere, and a newly build ferry wants to reach the waterway to enter it's service, it's route calculation seems to also include nodes where water and lava touches as possible routes, and then the ferry get's stuck in front of the lava, never reaching the destination waterway.

### To Reproduce
1. Open the attached savefile `bug_ferry_stuck_at_lava_and_no_mine_building_place_at_gold.wgf.zip` of a gameplay of the map `Long, long way v2.wmf`,
2. go to the place in the attached screenshot `ferry_stuck_at_lava.png`,
3. see that there is a ferry trying to reach the waterway build in the lower left side of the lake, but it is stuck at the lava.
   (By destroying the waterway the ferry stops it's aim to reach it. Re-building makes again a ferry trying to reach it; depending on the ferry's starting position the ferry succeeds or not.)

### Expected behaviour
The ferrys navigate around lava cleanly and do not get stuck.

### Screenshots
Attached: `ferry_stuck_at_lava.png`.

### Version:
* OS: [Artix GNU/Linux](https://artixlinux.org/), 64 Bit
* Widelands Version (output of `widelands --version`): `1.0~git24817[73e7f04@master] (Release)`
   (development version, after release 21, commit count 24817, git hash `73e7f04`).

---

#### Attached files:
* Savegame: [bug_ferry_stuck_at_lava_and_no_mine_building_place_at_gold.wgf.zip](https://github.com/widelands/widelands/files/5143341/bug_ferry_stuck_at_lava_and_no_mine_building_place_at_gold.wgf.zip "bug_ferry_stuck_at_lava_and_no_mine_building_place_at_gold.wgf.zip")
* Screenshot:

### LINKED ISSUE #4174: In Map 'Long, long way v2.wmf', there is no space to build a mine at gold where from symmetry it seems that there should be.
### Describe the bug
In the map `Long, long way v2.wmf`, there is a place where there is gold, and symmetry of the maps make me assume that there should be space to build a mine. But there is no space (only space for flags).

### To Reproduce
1. Open the attached savefile `bug_ferry_stuck_at_lava_and_no_mine_building_place_at_gold.wgf.zip` of a gameplay of the map `Long, long way v2.wmf`,
2. go to the place in the attached screenshot `no_mine_place_at_gold.png`,
3. see that there is gold but no mine can be build. Forther north, there is a similar place and a mine can be built.

### Expected behaviour
A mine building place at the spot.

### Screenshots
Attached: `ferry_stuck_at_lava.png`.

### Version:
* OS: [Artix GNU/Linux](https://artixlinux.org/), 64 Bit
* Widelands Version (output of `widelands --version`): `1.0~git24817[73e7f04@master] (Release)`

---

#### Attached files:
* Savegame: [bug_ferry_stuck_at_lava_and_no_mine_building_place_at_gold.wgf.zip](https://github.com/widelands/widelands/files/5143342/bug_ferry_stuck_at_lava_and_no_mine_building_place_at_gold.wgf.zip "bug_ferry_stuck_at_lava_and_no_mine_building_place_at_gold.wgf.zip")
* Screenshot:
  no_mine_place_at_gold.zip

### PR DESCRIPTION
**Pull request by Noordfrees**
Fixes #4173 : Don't allow ferries to swim on lava
Fixes #4174 : Re-add some mining plots to Long Long Way map

### CODE REVIEW

File: src/logic/map_objects/checkstep.cc [hunk 1]
```diff
@@ -121,7 +121,12 @@ CheckStepFerry
 */
 bool CheckStepFerry::allowed(
    const Map& map, const FCoords& from, const FCoords& to, int32_t dir, CheckStep::StepId) const {
+	if (!(to.field->nodecaps() & MOVECAPS_WALK) && !(to.field->nodecaps() & MOVECAPS_SWIM)) {
+		// can't swim on lava
+		return false;
+	}
 	if (MOVECAPS_SWIM & (from.field->nodecaps() | to.field->nodecaps())) {
```
Comment (gunchleoc) [inline src/logic/map_objects/checkstep.cc:128]: Do you need to check the `to.field->nodecaps()` here again?

**Review by gunchleoc — COMMENTED** [review]
Can you please describe the changes to the map in the PR comment?

File: src/logic/map_objects/checkstep.cc (hunk 1 above)
Comment (Noordfrees) [reply src/logic/map_objects/checkstep.cc:128]: The first `if` checks whether the destination field is near lava, in which case we know immediately that we mustn't go there. The second `if` checks whether either field is open sea, in which case we know the move is allowed. When we get to this `if`, we know that `to.field` has `MOVECAPS_WALK` *or* `MOVECAPS_SWIM`, but we still need to check which one.
Often, both fields will have `MOVECAPS_WALK` but neither has `MOVECAPS_SWIM` so we need to check at large with the `switch` below.

Comment (gunchleoc) [reply src/logic/map_objects/checkstep.cc:128]: Now I see it. Thanks for the explanation :)

**Review by hessenfarmer — APPROVED** [review]
Code LGTM
Checked map in editor

**Pull request merged**